src/iceberg/util/snapshot_util.cc

```diff
@@ -0,0 +1,373 @@
+  }
+
+  if (last_snapshot && !last_snapshot->parent_snapshot_id.has_value()) {
+    // this is the first snapshot in the table, return it
+    return last_snapshot;
+  }
+
+  return ValidationFailed("Cannot find snapshot older than {}",
+                          FormatTimestamp(timestamp_ms));
+}
+
+Result<std::vector<int64_t>> SnapshotUtil::SnapshotIdsBetween(const Table& table,
+                                                              int64_t from_snapshot_id,
+                                                              int64_t to_snapshot_id) {
+  ICEBERG_ASSIGN_OR_RAISE(auto to_snapshot, table.SnapshotById(to_snapshot_id));
+  if (!to_snapshot) {
+    return InvalidArgument("Cannot find snapshot: {}", to_snapshot_id);
+  }
+
+  // Create a lookup function that returns null when snapshot_id equals from_snapshot_id
+  // This effectively stops traversal at from_snapshot_id (exclusive)
+  auto lookup = [&table,
+                 from_snapshot_id](int64_t id) -> Result<std::shared_ptr<Snapshot>> {
+    if (id == from_snapshot_id) {
+      return nullptr;
+    }
+    return table.SnapshotById(id);
+  };
+
+  auto ancestors = AncestorsOf(to_snapshot, lookup);
+  return ToIds(ancestors);
+}
+
+Result<std::vector<int64_t>> SnapshotUtil::AncestorIdsBetween(
+    const Table& table, int64_t latest_snapshot_id,
+    const std::optional<int64_t>& oldest_snapshot_id) {
+  ICEBERG_ASSIGN_OR_RAISE(
+      auto ancestors, AncestorsBetween(table, latest_snapshot_id, oldest_snapshot_id));
+  return ToIds(ancestors);
+}
+
+Result<std::vector<std::shared_ptr<Snapshot>>> SnapshotUtil::AncestorsBetween(
+    const Table& table, int64_t latest_snapshot_id,
+    const std::optional<int64_t>& oldest_snapshot_id) {
+  ICEBERG_ASSIGN_OR_RAISE(auto start, table.SnapshotById(latest_snapshot_id));
+  if (!start) {
+    return InvalidArgument("Cannot find snapshot: {}", latest_snapshot_id);
+  }
+
+  if (oldest_snapshot_id.has_value()) {
+    if (latest_snapshot_id == oldest_snapshot_id.value()) {
+      return std::vector<std::shared_ptr<Snapshot>>();
+    }
+
+    auto lookup = [&table, oldest_snapshot_id = oldest_snapshot_id.value()](
+                      int64_t id) -> Result<std::shared_ptr<Snapshot>> {
+      if (id == oldest_snapshot_id) {
+        return nullptr;
+      }
+      return table.SnapshotById(id);
+    };
+    return AncestorsOf(start, lookup);
+  } else {
+    return AncestorsOf(table, start);
+  }
+}
+
+std::vector<std::shared_ptr<Snapshot>> SnapshotUtil::AncestorsOf(
+    const Table& table, const std::shared_ptr<Snapshot>& snapshot) {
+  std::vector<std::shared_ptr<Snapshot>> result;
+  if (!snapshot) {
+    return result;
+  }
+
+  std::shared_ptr<Snapshot> current = snapshot;
+  while (current) {
+    result.push_back(current);
+    if (!current->parent_snapshot_id.has_value()) {
+      break;
+    }
+    auto parent_result = table.SnapshotById(current->parent_snapshot_id.value());
+    if (!parent_result.has_value()) {
```

Comment:
We might need to check if the error is `NotFound` and return the original error for other kind of errors.